tools/funccount.py

```diff
@@ -57,18 +57,25 @@ def signal_ignore(signal, frame):
 bpf_text = """
 #include <uapi/linux/ptrace.h>
 
-struct key_t {
+#define MAP_SIZE 10000
+
+struct val_t {
     u64 ip;
+    u64 value;
 };
-BPF_HASH(counts, struct key_t);
+BPF_TABLE("array", int, struct val_t, counts, MAP_SIZE);
 
 int trace_count(struct pt_regs *ctx) {
     FILTER
-    struct key_t key = {};
-    u64 zero = 0, *val;
-    key.ip = ctx->ip;
-    val = counts.lookup_or_init(&key, &zero);
-    (*val)++;
+    struct val_t *val;
+    u64 zero = 0;
+    u32 key = ctx->ip % MAP_SIZE;
```

Comment:
@goldshtn right, so this update drops events if there was a collision:

```
    if (!val || (val->ip != 0 && val->ip != ctx->ip))
        return 0;
```

If, instead, it:

```
    if (!val || (val->ip != 0 && val->ip != ctx->ip))
        if (val->ip != ctx->ip)
            collisions++;
        return 0;
```
And then told the end-user the collisions counter if it was non-zero, then this solution is workable. If you see collisions, then match fewer functions. If there were no collisions, then no problem. This will probably have to do until the kernel supports recursive context hashs.